Added section for merging asset nodes

TSEcho opened the pull request:
Created an additional section in the AssetAPI docs for how to merge nodes from an existing Asset to a current configuration node. Have had several people ask for something like this and it seems to be better to set this here than in the configuration node section as chronologically the user may not have looked at assets yet which is required here.

Commit:
Added section for merging asset nodes 

Created an additional section in the AssetAPI docs for how to merge nodes from an existing Asset to a current configuration node. Have had several people ask for something like this and it seems to be better to set this here than in the configuration node section as chronologically the user may not have looked at assets yet which is required here.

Spongy commented:
A preview for this pull request is available at https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/3b0ac52/.

Here are some links to the pages that were modified:

- modified: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/3b0ac52/plugin/configuration/loaders.html

_Since the preview frequently changes, please link to [this comment](https://github.com/SpongePowered/SpongeDocs/pull/801#issuecomment-493345607), not to the direct url to the preview._

ST-DDT commented on source/plugin/assets.rst:
```diff
@@ -59,3 +59,16 @@ following code:
     within the SpongeAPI environment in favor of the ``AssetManager`` in order to provide a more confluent way of
     retrieving resources not only within your own plugin, but for other plugins as well.
+
+Working with Assets
```
This headline exists already.

TSEcho replied:
> IMO this section should be better moved to the config section. From the context it does not match here.
> 
> Maybe update this page:
> https://docs.spongepowered.org/stable/en/plugin/configuration/loaders.html?highlight=config#example-loading-a-default-config-from-the-plugin-jar-file

Yeah I can see It making sense that way. Put it in assets originally as assets explanation came after but that link shows assets are linked prior and an example is shown so they dev should understand by then. Newest commit will reflect those changes. And fixed that header issue; thought I had changed that earlier honestly

ST-DDT reviewed (changes requested):
IMO this section should be better moved to the config section. From the context it does not match here.

Maybe update this page:
https://docs.spongepowered.org/stable/en/plugin/configuration/loaders.html?highlight=config#example-loading-a-default-config-from-the-plugin-jar-file

Commit:
Removed previous commit section

will be adding it to under the default config area instead

Commit:
Moved merge node section to loaders from assets

Was recommended that this section be moved here for better categorization from my previous commit. Also added an extra note explaining how the nodes will not replace existing nodes but rather only set them up if they are virtual.

Commit:
Resolved merge conflict with title being too short

Commit:
Re: Fixed merge conflict

Apparently the viewer doesn't reflect how long the underline actually is so its been shorter the entire time

Commit:
Re:Re fixed merge conflict

I STG you should never try to fix things when you just wake up. This should fix it up now

ST-DDT commented:
The two sections at the end of the page look redundant, don't they? Can you merge them together? The old code example has been argued against in the past so you could replace it with yours. Or explain the differences that both yield.

ghost commented:
> The two sections at the end of the page look redundant, don't they? Can you merge them together? The old code example has been argued against in the past so you could replace it with yours. Or explain the differences that both yield.

Would say the default config section for loaders is a bit redundant considering its explicitly referring to using an asset. That and there is a good bit of unneeded code there as that entire segment could be done in one line via `Asset#copyToFile(asset, false, true). Might have to edit a few things

Commit:
Removed redundant code + added explanations

Previous load from default config was using the AssetAPI but still declared it as a separate URL and had a check to see if the map was empty in an if statement which is a bit more boilerplate than needed. Replaced that with the `copyToFile(path, false, true)` method so its easier to use. Also added explanations as to why you'd use mergeValuesFrom instead of simply copying them as the two have different uses.

ST-DDT commented on source/plugin/configuration/loaders.rst:
```diff
@@ -26,7 +26,7 @@ configurations, giving you the option of hard-coding default values or loading f
 
 Getting your Loader
-~~~~~~~~~~~~~~~~~~~
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Unneccesary change

ST-DDT commented on source/plugin/configuration/loaders.rst:
```diff
@@ -140,22 +140,17 @@ Example: Loading a Default Config from the Plugin Jar File
 must not lie in the jar file root, but instead in the directory ``assets/myplugin``. For more information, see
 :doc:`the Asset API page <../assets>`.
 
+Additionally, instead of checking to see if the node has an empty children map, this method uses ``copyToFile(path, false, true)``. The first argument is the path in which to set the node to, the second represents whether or not the file should be overriden each time, and the final argument represents if the existing config should be overriden only if absent.
```
Not sure about `additionally` from the context there . There is no check for an empty node here. Maybe rephrase it somewhat like: This variant should be used instead of checking whether the root node is empty.

Please keep the line length limit in mind.

ST-DDT commented on source/plugin/configuration/loaders.rst:
```diff
@@ -164,3 +159,22 @@ must not lie in the jar file root, but instead in the directory ``assets/myplugi
 If you have an extra configuration class, you can use a much easier approach that also works if the only a part of your
 config is missing. See also the examples on the :doc:`serialization` page.
+
+Merging Nodes from an Asset
```
Maybe add the `Example:`  prefix?

ST-DDT commented on source/plugin/configuration/loaders.rst:
```diff
@@ -164,3 +159,22 @@ must not lie in the jar file root, but instead in the directory ``assets/myplugi
 config is missing. See also the examples on the :doc:`serialization` page.
+
+Merging Nodes from an Asset
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
The headline marker should be as long as the headline.

ST-DDT commented on source/plugin/configuration/loaders.rst:
```diff
@@ -140,22 +140,17 @@ Example: Loading a Default Config from the Plugin Jar File
 
-    import java.net.URL;
-
+    Sponge.getAssetManager().getAsset(myplugin, "defaultConfig.conf").get().copyToFile(path, false, true);
```
Please use the same asset name everywhere. It does not matter which one you use.

ST-DDT reviewed (changes requested):
This is way better.

Commit:
Asset names matching change

Changed the names for the assets to all be default.conf

Commit:
Changed headline for syntax + added "Example"

Commit:
Wording changes + markup modifications

Removed any excessive headline markup and changed up the wording for why you should use copyToFile() instead of checking for an empty node.

ST-DDT reviewed (approved):
Looks good to me. 

It would be nice to get a short review from the API team as well.

dualspiral commented on source/plugin/configuration/loaders.rst:
```diff
@@ -140,22 +140,20 @@ Example: Loading a Default Config from the Plugin Jar File
 :doc:`the Asset API page <../assets>`.
 
+This variant should be used instead of checking whether the root node is empty as it simplifies the process.
+The first argument is the path in which to set the node, the second represents whether or not 
```
The first argument of what? This should be explicit in naming the `copyToFile` method - perhaps linking to it in the JDs and just explaining that the form will only copy the file to the `Path` if the file doesn't exist already.

TSEcho replied:
Added in next commit

dualspiral commented on source/plugin/configuration/loaders.rst:
```diff
@@ -140,22 +140,20 @@ Example: Loading a Default Config from the Plugin Jar File
 
```
I think some of the explanation of what this example does should go here - set up the context of the example then give the example. The example and code seem to just come out of nowhere.

Actually, I don't think this should be given an "example" moniker - perhaps change the title to "Creating a Default Configuration File".

> A popular way to provide a default configuration file with your plugin is to include a copy of it in your plugin jar, copying it to the config directory when the config file has yet to be created. You can use the Asset API (link?) to do this, as shown in the example below:

dualspiral commented on source/plugin/configuration/loaders.rst:
```diff
@@ -140,22 +140,20 @@ Example: Loading a Default Config from the Plugin Jar File
 must not lie in the jar file root, but instead in the directory ``assets/myplugin``. For more information, see
 :doc:`the Asset API page <../assets>`.
 
+This variant should be used instead of checking whether the root node is empty as it simplifies the process.
```
I think you should just drop this sentence - it adds nothing.

dualspiral commented on source/plugin/configuration/loaders.rst:
```diff
@@ -164,3 +162,23 @@ must not lie in the jar file root, but instead in the directory ``assets/myplugi
 If you have an extra configuration class, you can use a much easier approach that also works if the only a part of your
 config is missing. See also the examples on the :doc:`serialization` page.
+
+Example: Merging Nodes from an Asset
```
I'm probably going to receive the ire of @ST-DDT here... but this isn't an example, this is a thing you might want to do and we explain how to do them. The code is an example, perhaps, but the concept isn't and that concept is not an example of anything else displayed in the topic so far.

This is better entitled "Updating Configuration files from the Default Configuration" - or something to that effect, the title should explain the problem, "merging nodes from an asset" might just be a little obscure for some people.

ST-DDT replied:
I'm fine with it, as long as both sections use the prefix or both don't.

TSEcho replied:
Going with prefix gone for all of them. It seems to match the flow of the page better

dualspiral reviewed (commented):
Thanks for the submission! The biggest issue I forsee with this is the flow - it jumps around a little bit and certain parts need introducing or clarifying. I also disagree with the "example" moniker - each of these things are discrete concepts that have example code attached.

Code seems fine to me.

Commit:
Changed all DualSpiral Suggestions

Removed redundant line 
Changed things off of example to make entire sections explaining things
Specified that the asset must lie in resources/assets/myplugin as some have been confused where to put it before
Added link to JavaDocs for Asset#copyToFile()

phit commented on source/plugin/configuration/loaders.rst:
```diff
@@ -163,7 +162,7 @@ config should be overriden only if absent.
 
-Example: Merging Nodes from an Asset
+Updating Configuration Files from the Default Configuration
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
need to change underline here to match title length, that's why compile failed

TSEcho replied:
Ah crap. Thought I did

Commit:
Fixed syntax
